### PR TITLE
plugin: Replace readClusterState with existing watch events

### PR DESCRIPTION
**Pull request by sharnoff**
... and a few other commits tossed in for fun.

Probably best to review commit-by-commit.

The thing I'm most worried about is that this potentially may change buffer handling.

_Might_ be a good idea to go for #840 first, but I suspect that might also have its own issues.

This is a precursor to #517, to minimize the number of places we need to change resource handling.

### CODE REVIEW

**Comment by sharnoff** [comment]
> Do you mean these watch events? 

Broadly, watch events _like_ those - but it only relies on the Pod watch events, I think. (and specifically, this PR waits for only the initial Pod "add" events to be completely processed)

> I am trying to understand how this relates to #873.

Right, yeah fair. This PR should be ok - we only care about the Pod watch events. (unless I missed something - please feel free to double-check me no that)

File: pkg/plugin/state.go [hunk 1]
```diff
@@ -728,16 +738,33 @@ func (e *AutoscaleEnforcer) reserveResources(
 	}
 	newNodeReservedCPU := node.cpu.Reserved + ps.cpu.Reserved
 	newNodeReservedMem := node.mem.Reserved + ps.mem.Reserved
+	newNodeBufferCPU := node.cpu.Buffer + ps.cpu.Buffer
```
Comment (Omrigan) [inline pkg/plugin/state.go:741]: This looks like `node.*.Buffer` is only increasing. Or am I missing something?

File: pkg/plugin/state.go [hunk 2]
```diff
@@ -686,20 +686,30 @@ func (e *AutoscaleEnforcer) reserveResources(
...
-			Buffer:           0,
+			Reserved:         vmInfo.Max().Mem,
+			Buffer:           util.SaturatingSub(vmInfo.Max().Mem, vmInfo.Using().Mem),
 			CapacityPressure: 0,
 			Min:              vmInfo.Min().Mem,
 			Max:              vmInfo.Max().Mem,
 		}
+
+		migrating := util.TryPodOwnerVirtualMachineMigration(pod) != nil
+		if !vmInfo.Config.ScalingEnabled || migrating {
```
Comment (Omrigan) [inline pkg/plugin/state.go:707]: When Buffer values are zeroed if ScalingEnabled=true? Also, these two lines had a comment before, maybe bring it along?

File: pkg/plugin/plugin.go [hunk 3]
```diff
@@ -118,15 +122,50 @@ func makeAutoscaleEnforcerPlugin(
...
+	//
+	// 1. Using watch.WatchModeSync to force queueing events *before* returning from creating the
+	//    watcher (note: and therefore, before any start to be handled); and
+	// 2. For each event created from the initial List(), increment a counter to track the number of
+	//    these events, and decrement it as events are handled.
+	//
+	// The initial state building is complete when the counter reaches zero, at which point we close
+	// the channel that this function will wait on.
+
+	var initEventsCount atomic.Int32
```
Comment (Omrigan) [inline pkg/plugin/plugin.go:140]: Can you use [semaphore](https://pkg.go.dev/golang.org/x/sync/semaphore#NewWeighted) here instead? 
 - Acquire(nil, n) when events are added
 - Release(1) when event is processed
 - Acquire(ctxWithTimeout, n) when waiting for events processing

File: pkg/plugin/plugin.go [hunk 4]
```diff
@@ -118,15 +122,50 @@ func makeAutoscaleEnforcerPlugin(
...
 	nwc := nodeWatchCallbacks{
 		submitNodeDeletion: func(logger *zap.Logger, nodeName string) {
 			pushToQueue(logger, nodeName, func() { p.handleNodeDeletion(hlogger, nodeName) })
 		},
 	}
 	pwc := podWatchCallbacks{
-		submitStarted: func(logger *zap.Logger, pod *corev1.Pod) {
-			pushToQueue(logger, pod.Name, func() { p.handleStarted(hlogger, pod) })
+		submitStarted: func(logger *zap.Logger, pod *corev1.Pod, preexisting bool) {
+			incEventCountIfInit(preexisting)
```
Comment (Omrigan) [inline pkg/plugin/plugin.go:164]: Why passing bool instead of an if here?

File: pkg/plugin/config.go [hunk 5]
```diff
@@ -30,6 +30,14 @@ type Config struct {
 	// event queue.
 	EventQueueWorkers int `json:"eventQueueWorkers"`
 
+	// StartupEventHandlingTimeoutSeconds gives the maximum duration, in seconds, that we are
+	// allowed to wait to finish handling all of the initial events generated by reading the cluster
+	// state on startup.
+	//
+	// If event processing takes longer than this time, then plugin creation will fail, and the
+	// scheduler pod will retry.
+	StartupEventHandlingTimeoutSeconds int `json:"startupEventHandlingTimeoutSeconds"`
```
Comment (Omrigan) [inline pkg/plugin/config.go:39]: What does this timeout protects us against? 

If we have a lot of events on the start, maybe it is better to anyway process all of them? Otherwise, next initialization attempt will fail as well and plugin cannot start up. 

We can also not fail plugin's pod, and just keep on with inaccurate scheduling.

**Review by Omrigan — COMMENTED** [review]
Have some questions

File: pkg/plugin/config.go (hunk 5 above)
Comment (sharnoff) [reply pkg/plugin/config.go:39]: The intent is that this should protect us against deadlocks, so that if event processing doesn't make enough process, we just restart.

File: pkg/plugin/state.go (hunk 1 above)
Comment (sharnoff) [reply pkg/plugin/state.go:741]: During reserve, yeah, that's correct.

File: pkg/plugin/state.go (hunk 2 above)
Comment (sharnoff) [reply pkg/plugin/state.go:709]: > When Buffer values are zeroed if ScalingEnabled=true?

Not sure what you're asking, could you rephrase?

> Also, these two lines had a comment before, maybe bring it along?

Ah good catch, thanks! Fixed in b3f2d70a764418b44ce502b81e496cf1b5084322

File: pkg/plugin/plugin.go (hunk 3 above)
Comment (sharnoff) [reply pkg/plugin/plugin.go]: Seems simpler. Implemented it in caba29f33339c53e094935e66239ee375d01b26a but tbh I'm not sure if it's better. WDYT?

File: pkg/plugin/plugin.go (hunk 4 above)
Comment (sharnoff) [reply pkg/plugin/plugin.go]: Initially it was because I thought there'd be more than one callback that would be involved here, and that would keep callback functions small.

It's now clear that's not necessary though, so moved the if here in 564ec88212cc6571e3620737a1d48450850b2174

File: pkg/plugin/config.go (hunk 5 above)
Comment (Omrigan) [reply pkg/plugin/config.go:39]: And how does it help? In case of a restart it will just keep trying to process the same events over and over again. 

But, if the event processing is not deadlocked, but just slow, it will take more time, but eventually the situation gets fixed.

Maybe instead of a restart it is better to fire an alert?

File: pkg/plugin/plugin.go (hunk 3 above)
Comment (Omrigan) [reply pkg/plugin/plugin.go]: If the `incEventCount` is left as a semaphore wrapper anyway, maybe instead revert to your old implementation, but wrap all in a struct synchronization primitive?

The reason I suggested semaphore is that I am wary of using raw channels and atomics in the code, where it feels like it should be a synchronization primitive.

File: pkg/plugin/state.go (hunk 2 above)
Comment (Omrigan) [reply pkg/plugin/state.go:709]: > > When Buffer values are zeroed if ScalingEnabled=true?
> 
> Not sure what you're asking, could you rephrase?

If ScalingEnabled=true, then buffer values are non-zero. How do they become zero? When we hear the report from the autoscaler-agent?

File: pkg/plugin/state.go (hunk 1 above)
Comment (Omrigan) [reply pkg/plugin/state.go:743]: How do they decrease? I guess the answer is the same as for the above comment.

File: pkg/plugin/plugin.go (hunk 3 above)
Comment (sharnoff) [reply pkg/plugin/plugin.go:140]: Made it a separate type in 521597e44affe70a50cccd1e0bf72d46fa500394 - LMK what you think.

File: pkg/plugin/state.go (hunk 2 above)
Comment (sharnoff) [reply pkg/plugin/state.go:709]: Correct, yeah - when handling the first autoscaler-agent request.

File: pkg/plugin/config.go (hunk 5 above)
Comment (sharnoff) [reply pkg/plugin/config.go:39]: I don't think we currently have a way for components to directly trigger an alert, at least not without metrics getting scraped. If the scheduler is down, someone will be paged either way - and we already have some alerts for if it's crashlooping.

I expect that restarting _may_ help against deadlocks because I _think_ it's more likely for us to have inconsistent deadlocks than it is for initial event processsing to just be really slow. But that's just a guess.

File: pkg/plugin/plugin.go (hunk 3 above)
Comment (Omrigan) [reply pkg/plugin/plugin.go:140]: I imagined the implementation with initEventsCount hidden inside the new helper.

Also, you don't need incEventCount now

File: pkg/plugin/config.go (hunk 5 above)
Comment (Omrigan) [reply pkg/plugin/config.go:39]: > I don't think we currently have a way for components to directly trigger an alert

I meant, expose a metric startup_event_handling_seconds, and configure an alert when it is >30s.

> I expect that restarting may help against deadlocks because I think it's more likely for us to have inconsistent deadlocks than it is for initial event processsing to just be really slow. But that's just a guess.

If we were to have an alert:
 - If it is a deadlock, then by logs it should be possible to see if event processing is progressing or not, and a scheduler can be restarted manually
 - If it is slow, then the issue improves on it's own. 

Also, deadlock means a bug in the code, and writing specific protections for bugs seems like a grey path (why not just fix the bug?). 

But I guess we don't actually expect any of the issues to be happening regularly, and even we observe too many events, the mitigation is easy -> bump the timeout. So, leaving this up to you at this point.

**Review by Omrigan — APPROVED** [review]

File: pkg/plugin/plugin.go (hunk 3 above)
Comment (sharnoff) [reply pkg/plugin/plugin.go:140]: > I imagined the implementation with initEventsCount hidden inside the new helper.

Hm, it's hard to make that kind of abstraction safe to use (because the counter isn't monotonic) -- maybe it'd be fine here, but I'd rather keep the logic inlined than make an unsafe abstraction for just this purpose.

So in its current state, incEventCount or equivalent is still needed (we can just directly add to the atomic; don't know whether that's better)

File: pkg/plugin/config.go (hunk 5 above)
Comment (sharnoff) [reply pkg/plugin/config.go:39]: > expose a metric startup_event_handling_seconds

The tricky thing with this kind of metric is that we want it to keep updating even while we're waiting. AFAICT that's non-trivial to implement, so will keep everything as is for now, but I'd be happy to revisit this at some point in the future.

**Comment by sharnoff** [comment]
Realized that the necessity of debe3b22e89af228af757094cad22bb26f316b80 (from #936) is partially due to 25bbe43ce8163d0290cbfb210c3b1a2eb13cb866 here — it adds Buffer always, but there's a handful of circumstances where we don't actually need it (like when a new VM is added but doesn't go through our scheduler)

**Comment by sharnoff** [comment]
Adapted debe3b22e89af228af757094cad22bb26f316b80 here as c8e250e4ebc0130e001cdd5f196af26ad209ffe7 to fix that issue. Impact should have been close to zero in practice, but wanted to fix it just to be safe. @Omrigan wdyt?

File: pkg/plugin/state.go [hunk 6]
```diff
@@ -585,7 +585,18 @@ func (e *AutoscaleEnforcer) handleStarted(logger *zap.Logger, pod *corev1.Pod) {
 
 	logger.Info("Handling Pod start event")
 
-	_, _, _ = e.reserveResources(context.TODO(), logger, pod, "Pod started", false)
+	_, _, _ = e.reserveResources(context.TODO(), logger, pod, "Pod started", reserveOptions{
+		// pod already started, out of our control - we don't have a mechanism to deny it
+		allowDeny: false,
+		// this may be a preexisting VM. If so, we should include it in "buffer" as long it's
+		// supposed to be handled by us (otherwise, the "buffer" will never be resolved)
+		includeBuffer: pod.Spec.SchedulerName == e.state.conf.SchedulerName,
```
Comment (Omrigan) [inline pkg/plugin/state.go:593]: Why do we even have SchedulerName as a part of the spec, why do we ever want it to be different?

**Review by Omrigan — APPROVED** [review]

**Comment by Omrigan** [comment]
Are you going to do rebase merge? If so, tmp commit needs to be squashed somewhere, maybe into middle commit.

File: pkg/plugin/state.go (hunk 6 above)
Comment (sharnoff) [reply pkg/plugin/state.go:593]: Result of discussing this: it's part of the pod spec because neonvm is a logically separate dependency of "autoscaling", so the ability to run with/without the rest of autoscaling requires the ability to set the pod's scheduler name
